[Rework]: Array (#704)
exercism/swift: exercises/concept/magician-in-training/.docs/introduction.md

# About

[Arrays][array] are one of Swift's three primary collection types.
Arrays are ordered lists of elements where the elements can be of any type, however, all elements of any given list must have the same type.
Arrays are mutable when assigned to a variable, meaning that the elements of an array can be modified after the array is created.
This is not the case when an array is assigned to a constant, in which case the array is immutable.

> Arrays literals are written as a series of elements, each separated by commas, enclosed in square brackets.

Reviewer: ```suggestion
Array literals are written as a series of elements, each separated by a comma, enclosed in square brackets.
```

> Swift will infer the type of the array from the types of the elements in the array literal.

Reviewer: ```suggestion
Swift will infer the type of the array from the type of the elements in the array literal.
```

```

## Add an Array to an Array

Array can be added to the end of an array using the `+` operator.
It is important to note that the `+` operator creates a new array and does not modify the original array, which is different from the `append(_:)` or `insert(_:at:)` methods.

```swift
var oddInts = [1, 3, 5, 7, 9, 11, 13]
oddInts + [15, 17, 19]
// returns [1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

print(oddInts)
// prints [1, 3, 5, 7, 9, 11, 13]
```

## Accessing Elements of an Array

Elements of an array can be accessed individually by supplying the index of the element inside square brackets following the array.
The index of an element is an `Int` and starts with `0` for the first (leftmost) element.
If the index is outside the valid range of indices, a runtime error will occur and the program will crash.

```swift
let evenInts = [2, 4, 6, 8, 10, 12]
let oddInts = [1, 3, 5, 7, 9, 11, 13]

evenInts[2]
// returns 6

oddInts[7]
// Fatal error: Index out of range
```

## Modifying Elements of an Array

Elements of an array can be modified by assigning a new value to the element at a given index.
The index of an element is an `Int` and starts with `0` for the first (leftmost) element.

> If the index is outside the valid range of indices, a runtime error will occur and the program will crash.

Reviewer: ```suggestion
If the given index is outside the valid range of indices for the array, a runtime error will occur and the program will crash.
```

evenInts[2] = 0
// evenInts is now [2, 4, 0, 8, 10, 12]
```

## Converting an Array to a String and Back

An array of `n` strings can be converted to a single string using the [`joined(separator:)`][joined] method.
The `joined(separator:)` method takes a single argument, the separator to be used between elements of the array.
The separator must be a string.

```swift
let evenInts = ["2", "4", "6", "8", "10", "12"]
let evenIntsString = evenInts.joined(separator: ", ")
// returns "2, 4, 6, 8, 10, 12"
```

An array can be converted from a string using the [`split(separator:)`][split] method.
The `split(separator:)` method takes a single argument, the separator to be used between elements of the array.

```swift
let evenIntsString = "2, 4, 6, 8, 10, 12"
let evenInts = evenIntsString.split(separator: ", ")
// returns ["2", "4", "6", "8", "10", "12"]
```

## Delete an Element from an Array

Elements can be deleted from an array using the [`remove(at:)`][remove] method.
The `remove(at:)` method takes a single argument, the index of the element to be removed from the array.
The index of an element is an `Int` and starts with `0` for the first (leftmost) element.

> If the index is outside the valid range of indices, a runtime error will occur and the program will crash.

Reviewer: ```suggestion
If the given index is outside the valid range of indices for the array, a runtime error will occur and the program will crash.
```
